pallets/pallet-bonded-coins/src/curves/mod.rs

```diff
@@ -0,0 +1,120 @@
+pub(crate) mod lmsr;
+pub(crate) mod polynomial;
+pub(crate) mod square_root;
+
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_arithmetic::ArithmeticError;
+use sp_std::ops::{AddAssign, BitOrAssign, ShlAssign};
+use substrate_fixed::traits::{Fixed, FixedSigned, ToFixed};
+
+use crate::{
+	curves::{
+		lmsr::{LMSRCalculation, LMSRFunctionParameters},
+		polynomial::PolynomialParameters,
+		square_root::SquareRootParameters,
+	},
+	PassiveSupply, Precision,
+};
+
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub enum Curve<Parameter> {
+	Polynomial(PolynomialParameters<Parameter>),
+	SquareRoot(SquareRootParameters<Parameter>),
+	LMSR(LMSRFunctionParameters<Parameter>),
+}
+
+pub enum Operation<PassiveSupply> {
+	Mint(PassiveSupply),
+	Burn(PassiveSupply),
+}
+
+impl<Balance> Operation<PassiveSupply<Balance>> {
+	pub fn inner_value(&self) -> &PassiveSupply<Balance> {
+		match self {
+			Operation::Mint(x) => x,
+			Operation::Burn(x) => x,
+		}
+	}
+}
+
+impl<Parameter> Curve<Parameter>
+where
+	Parameter: FixedSigned + PartialOrd<Precision> + From<Precision> + ToFixed,
+	<Parameter as Fixed>::Bits: Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign,
+{
+	pub fn calculate_cost(
+		&self,
+		active_issuance_pre: Parameter,
+		active_issuance_post: Parameter,
+		op: Operation<PassiveSupply<Parameter>>,
+	) -> Result<Parameter, ArithmeticError> {
+		match self {
+			Curve::Polynomial(params) => {
+				let (low, high) = calculate_integral_bounds(op, active_issuance_pre, active_issuance_post);
+
+				params.calculate_costs(low, high)
```

Comment:
if all curves used the same trait (such as the signature of the calculate_cost function defined here) you could probably just pass down all arguments with a match all pattern

Reply:
Like so: [696c7b8](https://github.com/KILTprotocol/kilt-node/pull/765/commits/696c7b898863d201e411e1f42b6c84b3f3476497)

Reply:
I also noticed a little bug while updating the trait bounds: [c55b6a9](https://github.com/KILTprotocol/kilt-node/pull/765/commits/c55b6a941d6cfe0021e620a78fb2ee4734c271a9)

Reply:
so much better! 
```rust
		match self {
 			Curve::Polynomial(params) => params.calculate_costs(active_issuance_pre, active_issuance_post, op),
 			Curve::SquareRoot(params) => params.calculate_costs(active_issuance_pre, active_issuance_post, op),
 			Curve::LMSR(params) => params.calculate_costs(active_issuance_pre, active_issuance_post, op),
 		}
```
still seems so weird to me though, is there really no pattern that avoids this repetition?

Reply:
At least, I couldn't find a smooth way. Was experimenting a lot on this topic. I could use generics, but I found it more confusing.